cli/loader_lib.c

```diff
@@ -0,0 +1,194 @@
+#endif
+static char dep_libs[256] = DEP_LIBS;
+
+void print_stderr(const char * msg)
+{
+    fputs(msg, stderr);
+}
+// I use three arguments a lot.
+void print_stderr3(const char * msg1, const char * msg2, const char * msg3)
+{
+    print_stderr(msg1);
+    print_stderr(msg2);
+    print_stderr(msg3);
+}
+
+
+/* Absolute path to the path of the current executable, gets filled in by `get_exe_path()` */
+static void * load_library(const char * rel_path, const char * src_dir) {
+    char path[2*PATH_MAX + 1] = {0};
+    strncat(path, src_dir, sizeof(path) - 1);
+    strncat(path, PATHSEPSTRING, sizeof(path) - 1);
+    strncat(path, rel_path, sizeof(path) - 1);
+
+    void * handle = NULL;
+#if defined(_OS_WINDOWS_)
+    wchar_t wpath[2*PATH_MAX + 1] = {0};
+    if (!utf8_to_wchar(path, wpath, 2*PATH_MAX)) {
+        print_stderr3("ERROR: Unable to convert path ", path, " to wide string!\n");
+        exit(1);
+    }
+    handle = (void *)LoadLibraryExW(wpath, NULL, LOAD_WITH_ALTERED_SEARCH_PATH);
+#else
+    handle = dlopen(path, RTLD_NOW | RTLD_GLOBAL);
+#endif
+
+    if (handle == NULL) {
+        print_stderr3("ERROR: Unable to load dependent library ", path, "\n");
+#if defined(_OS_WINDOWS_)
+        LPWSTR wmsg = TEXT("");
+        FormatMessageW(FORMAT_MESSAGE_ALLOCATE_BUFFER |
+                       FORMAT_MESSAGE_FROM_SYSTEM |
+                       FORMAT_MESSAGE_IGNORE_INSERTS |
+                       FORMAT_MESSAGE_MAX_WIDTH_MASK,
+                       NULL, GetLastError(),
+                       MAKELANGID(LANG_ENGLISH, SUBLANG_ENGLISH_US),
+                       (LPWSTR)&wmsg, 0, NULL);
+        char err[256] = {0};
+        wchar_to_utf8(wmsg, err, 255);
+        print_stderr3("Message:", err, "\n");
+#else
+        print_stderr3("Message:", dlerror(), "\n");
+#endif
+        exit(1);
+    }
+    return handle;
+}
+
+char exe_dir[PATH_MAX];
+const char * get_exe_dir()
+{
+#if defined(_OS_WINDOWS_)
+    // On Windows, we use GetModuleFileName()
+    wchar_t julia_path[PATH_MAX];
+    if (!GetModuleFileName(NULL, julia_path, PATH_MAX)) {
+        print_stderr("ERROR: GetModuleFileName() failed\n");
+        exit(1);
+    }
+    if (!wchar_to_utf8(julia_path, exe_dir, PATH_MAX)) {
+        print_stderr("ERROR: Unable to convert julia path to UTF-8\n");
+        exit(1);
+    }
+#elif defined(_OS_DARWIN_)
+    // On MacOS, we use _NSGetExecutablePath(), followed by realpath()
+    char nonreal_exe_path[PATH_MAX + 1];
+    uint32_t exe_path_len = PATH_MAX;
+    int ret = _NSGetExecutablePath(nonreal_exe_path, &exe_path_len);
+    if (ret != 0) {
+        print_stderr("ERROR: _NSGetExecutablePath() failed\n");
+        exit(1);
+    }
+
+    if (realpath(nonreal_exe_path, exe_dir) == NULL) {
+        print_stderr("ERROR: realpath() failed\n");
+        exit(1);
+    }
+#elif defined(_OS_LINUX_)
+    // On Linux, we read from /proc/self/exe
+    int num_bytes = readlink("/proc/self/exe", exe_dir, PATH_MAX);
```

Comment:
Also, this should be fixed. I have definately used julia in chroot environments that does not `/proc` mounted.
It was usually a mistake of course but it was previously working.

Reply:
Do you have a good way to find the location of the current process without using `/proc`?

Reply:
No, but the system linker behavior is to fallback to default.

Reply:
And again, this should just be linked in.